packages/flutter_tools/lib/src/device.dart

```diff
@@ -61,9 +61,11 @@ class DeviceManager {
 
   /// Return the list of connected devices, filtered by any user-specified device id.
   Stream<Device> getDevices() {
-    return hasSpecifiedDeviceId
-        ? getDevicesById(specifiedDeviceId)
-        : getAllConnectedDevices();
+    if (specifiedDeviceId == null || specifiedDeviceId == 'all') {
```

Comment:
Changed so that `hasSpecifiedDeviceId` is true only if a single device is specified, `specifiedDeviceId` is null for -d all, and added `hasSpecifiedAllDevices`. After this fix, logic around `hasSpecifiedDeviceId` is as before.